Add the culprit json fragment in the error messages systematically.

EmileTrotignon opened the pull request:
The culprit fragment is the piece of json that is wrong. It is printed with a special shallow printer, in order to keep the messages short and to the point.

I have included new tests.

A bit of code smell is the fact that my code relies on the fact that the current json being converted is always called `x`. Help is appreciated on that front, in a lot of cases I could not get the proper expression.

I will probably add a bit more tests tomorrow.

davesnx commented on ppx/test/errors.t/run.t:
```diff
@@ -0,0 +1,19 @@
+    "d": [ 123, [ 1.2, 2.3, 2.4 ], "i am here" ]
+  }
+  $ dune exec ./prettify.exe -- tag_as_string.json
+  Fatal error: exception  exception Ppx_deriving_json_runtime.Of_json_error(Json_error "expected [\"A\"] or [\"Foo\"] or [\"B\", _] or [\"C\", _, _] or [\"D\", { _ }]. Json received: \"A\"")
```
I think this is 10x better than before, congrats! do you think adding the key would be even better? 

`expected blablalb at key "xxx"` or some other wording

EmileTrotignon replied:
Maybe. I think the "proper" way to do it could be to give a loc in the input file, because there won't always be a key (the variant might be in a list for instance). But getting locs is probably impossible (the yojson datastructure does not have them). So I can try and add more context in some places.

anmonteiro replied:
I'd also replace `or` with `|` to resemble the OCaml compiler's error messages for warning 8.

davesnx replied:
right, there may be cases where the key isn't on the level where you are decoding but still doable. If it gets messy, can always push another PR with it, don't want to block anything here

EmileTrotignon replied:
> I'd also replace or with | to resemble the OCaml compiler's error messages for warning 8.

I think `or` is more appropriate here because `|` is not part of the json syntax, and the errors are expressed in json syntax.

davesnx reviewed (approved):
🤘🏼

anmonteiro reviewed (approved):
this looks pretty great, nice work!

andreypopp commented on ppx/browser/ppx_deriving_json_js.ml:
```diff
@@ -65,13 +65,13 @@ module Of_json = struct
     [%expr
       if Stdlib.( <> ) [%e len] [%e eint ~loc n] then
-        Ppx_deriving_json_runtime.of_json_error
+        Ppx_deriving_json_runtime.of_json_msg_error ~json:x
```
```suggestion
        Ppx_deriving_json_runtime.of_json_msg_error ~json:[%e x]
```
let's pass `x` to this function so we can use like this ^

EmileTrotignon replied:
Its not that simple. I tried and I could not do it. The reason is that this function is called from `derive_of_variant_case` which does not have `x`, and `derive_of_variant_case` is directly given to `deriving_of` which uses objects and that I did not really understand. If you can find a way to make this cleaner it would be great.

andreypopp replied:
I was referring to the immediate call site 
```
[%e ensure_json_array_len ~loc 2 [%expr len] [%expr x]];
```
so that the assumptions about what names are defined do not spread to helper functions

EmileTrotignon replied:
okay why not.

EmileTrotignon replied:
This is done

andreypopp commented on ppx/browser/ppx_deriving_json_runtime.ml:
```diff
@@ -6,6 +6,100 @@ let to_string t = Js.Json.stringify t
+                Buffer.add_string buffer {|": |};
+                loop ?depth v;
+                Buffer.add_string buffer {|, |})
+              assoc;
```
should we limit the number of keys here? otherwise we can still output a lot

EmileTrotignon replied:
This is done

andreypopp commented on ppx/browser/ppx_deriving_json_runtime.ml:
```diff
@@ -6,6 +6,100 @@ let to_string t = Js.Json.stringify t
+              (fun elt ->
+                loop ?depth elt;
+                Buffer.add_string buffer ", ")
+              li;
```
same question: should we limit the number of elements and add `...` if there's more?

EmileTrotignon replied:
This is done

andreypopp commented on ppx/browser/ppx_deriving_json_runtime.ml:
```diff
@@ -6,6 +6,100 @@ let to_string t = Js.Json.stringify t
+        | `Null -> Buffer.add_string buffer "null"
+        | `String str ->
+            Buffer.add_string buffer {|"|};
+            Buffer.add_string buffer (String.escaped str);
```
I'd truncate big strings and add `...`

EmileTrotignon replied:
This is done

andreypopp commented on ppx/native/ppx_deriving_json_runtime.ml:
```diff
@@ -25,9 +40,59 @@ let show_json_type = function
   | `Null -> "null"
   | `String _ -> "string"
 
+let show_json_error ?depth json =
```
same comments about truncation of large objects/lists/strings here

andreypopp replied:
actually, given that we have `classify` in both browser and native runtimes — can we use same `show_json_error` for both browser and native?

EmileTrotignon replied:
Maybe I can attempt it yes. I liked the native code with Format.fprintf a bit better than the browser one with a buffer but I wrote both anyway.

EmileTrotignon replied:
This is done.

andreypopp commented on ppx/native/ppx_deriving_json_runtime.ml:
```diff
@@ -25,9 +40,59 @@ let show_json_type = function
+
+let of_json_error_mismatch ~depth ~expected ~json =
+  of_json_msg_error
+    ("expected "
+    ^ (expected |> List.map show_json_error |> String.concat "or ")
+    ^ "but got"
+    ^ show_json_error ~depth json)
```
use `Printf.sprintf` (ok in native)?

andreypopp commented on ppx/native/ppx_deriving_json_runtime.ml:
```diff
@@ -25,9 +40,59 @@ let show_json_type = function
-  of_json_error
-    ("expected " ^ expected ^ " but got " ^ show_json_type json)
+  of_json_msg_error
+    ("expected "
+    ^ expected
+    ^ " but got "
+    ^ show_json_type json
+    ^ ": "
+    ^ show_json_error ~depth:1 json)
```
`Printf.sprintf`?

andreypopp commented on ppx/browser/ppx_deriving_json_runtime.ml:
```diff
@@ -6,6 +6,100 @@ let to_string t = Js.Json.stringify t
+
+let of_json_error ?(depth = 2) ~json msg =
+  of_json_msg_error
+    (Printf.sprintf "%s but got %s" msg (show_json_error ~depth json))
```
we need to get rid of `Printf.sprintf` in browser runtime (I know it's used now but I got rid of it in #45)

andreypopp reviewed (changes requested):
This is a very good improvement!

Left some comments below.

andreypopp reviewed (approved):
Thank you!

EmileTrotignon commented:
I thought it was a bit confusing to have all the runtime files mixed with the ppx files, especially because I added a bunch, so I moved them to a separate folder.

If there is no issue with that, I have addressed every comment I got, I think its time to merge (I do not have the rights to do it).

jchavarri commented on ppx/browser/ppx_deriving_json_js.ml:
```diff
@@ -65,13 +65,13 @@ module Of_json = struct
     [%expr
       if Stdlib.( <> ) [%e len] [%e eint ~loc n] then
-        Ppx_deriving_json_runtime.of_json_error
+        Ppx_deriving_json_runtime.of_json_msg_error ~json:[%e x]
```
@EmileTrotignon This function doesn't take a `json` arg but somehow the tests passed. 🤔

jchavarri replied:
fixed in https://github.com/melange-community/melange-json/pull/45/commits/3fa710dc65665e8541f4aa72d648af5dcb990f95 and https://github.com/melange-community/melange-json/pull/45/commits/be46a4d83110deb341f4b53ca4d1d4b37fc24e41
